lib/virtio/virtio.c

```diff
@@ -96,22 +96,19 @@ int virtio_create_virtqueues(struct virtio_device *vdev, unsigned int flags,
 	/* Initialize virtqueue for each vring */
 	for (i = 0; i < nvqs; i++) {
 		vring_info = &vdev->vrings_info[i];
-		vring_alloc.vaddr = vring_info->va;
-		vring_alloc.align = vring_info->align;
-		vring_alloc.num_descs = vring_info->num_descs;
-
+		
+		vring_alloc = &vring_info->info;
 		if (vdev->role == VIRTIO_DEV_MASTER) {
 			size_t offset;
 			struct metal_io_region *io = vring_info->io;
-			unsigned int num_descs = vring_info->num_descs;
-			unsigned int align = vring_info->align;
 
-			offset = metal_io_virt_to_offset(io, vring_info->va);
+			offset = metal_io_virt_to_offset(io, vring_alloc->vaddr);
 			metal_io_block_set(io, offset, 0,
-					   vring_size(num_descs, align));
+					   vring_size(vring_alloc->num_descs,
+						      vring_alloc->align));
 		}
 		ret = virtqueue_create(vdev, i,
```

Comment:
how about just pass virtio_vring_info to virtqueue_create? and remove the requirement of vring_alloc_info?